backend/remote-state/etcdv3/client.go

```diff
@@ -0,0 +1,195 @@
+	Key    string
+
+	etcdMutex   *etcdv3sync.Mutex
+	etcdSession *etcdv3sync.Session
+	info        *state.LockInfo
+	mu          sync.Mutex
+	modRevision int64
+}
+
+func (c *RemoteClient) Get() (*remote.Payload, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	res, err := c.Client.KV.Get(context.TODO(), c.Key)
+	if err != nil {
+		return nil, err
+	}
+	if res.Count == 0 {
+		return nil, nil
+	}
+	if res.Count >= 2 {
+		return nil, fmt.Errorf("Expected a single result but got %d.", res.Count)
+	}
+
+	c.modRevision = res.Kvs[0].ModRevision
+
+	payload := res.Kvs[0].Value
+	md5 := md5.Sum(payload)
+
+	return &remote.Payload{
+		Data: payload,
+		MD5:  md5[:],
+	}, nil
+}
+
+func (c *RemoteClient) Put(data []byte) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	res, err := etcdv3.NewKV(c.Client).Txn(context.TODO()).If(
+		etcdv3.Compare(etcdv3.ModRevision(c.Key), "=", c.modRevision),
+	).Then(
+		etcdv3.OpPut(c.Key, string(data)),
+		etcdv3.OpGet(c.Key),
+	).Commit()
+
+	if err != nil {
+		return err
+	}
+	if !res.Succeeded {
+		return fmt.Errorf("The transaction did not succeed.")
+	}
+	if len(res.Responses) != 2 {
+		return fmt.Errorf("Expected two responses but got %d.", len(res.Responses))
+	}
+
+	c.modRevision = res.Responses[1].GetResponseRange().Kvs[0].ModRevision
+	return nil
+}
+
+func (c *RemoteClient) Delete() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	_, err := c.Client.KV.Delete(context.TODO(), c.Key)
+	return err
+}
+
+func (c *RemoteClient) Lock(info *state.LockInfo) (string, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if !c.DoLock {
+		return "", nil
+	}
+
+	c.info = info
+	return c.lock()
+}
+
+func (c *RemoteClient) Unlock(id string) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if !c.DoLock {
+		return nil
+	}
+
+	return c.unlock(id)
+}
+
+func (c *RemoteClient) getLockInfo() (*state.LockInfo, error) {
+	res, err := c.Client.KV.Get(context.TODO(), c.Key+lockInfoSuffix)
+	if err != nil {
+		return nil, err
+	}
+	if res.Count == 0 {
+		return nil, nil
+	}
+
+	li := &state.LockInfo{}
+	err = json.Unmarshal(res.Kvs[0].Value, li)
+	if err != nil {
+		return nil, fmt.Errorf("Error unmarshaling lock info: %s.", err)
+	}
+
+	return li, nil
+}
+
+func (c *RemoteClient) putLockInfo(info *state.LockInfo) error {
+	c.info.Path = c.Key
+	c.info.Created = time.Now().UTC()
+
+	_, err := c.Client.KV.Put(context.TODO(), c.Key+lockInfoSuffix, string(c.info.Marshal()))
+	return err
+}
+
+func (c *RemoteClient) lock() (string, error) {
+	session, err := etcdv3sync.NewSession(c.Client)
+	if err != nil {
+		return "", nil
+	}
+
+	ctx, cancel := context.WithTimeout(context.TODO(), lockAcquireTimeout)
+	defer cancel()
+
+	mutex := etcdv3sync.NewMutex(session, c.Key+lockSuffix)
+	if err1 := mutex.Lock(ctx); err1 != nil {
+		lockInfo, err2 := c.getLockInfo()
+		if err2 != nil {
+			return "", &state.LockError{Err: err2}
+		}
+		return "", &state.LockError{Info: lockInfo, Err: err1}
+	}
+
+	c.etcdMutex = mutex
+	c.etcdSession = session
```

Comment:
Again I'm not sure about etcd, but consul sessions favored liveness, and were easier to lose than desired and susceptible to network timeouts. I'm not sure if we need to do the same here, but the consul backend watches the session status and reconnects when necessary.

Reply:
I believe that's not the case with`etcd`, judging by a number of factors including this

https://github.com/coreos/etcd/blob/master/clientv3/concurrency/session.go#L26
https://github.com/coreos/etcd/blob/master/clientv3/concurrency/session.go#L64

and some tests I ran locally using `etcdctl`:

```
$ ETCDCTL_API=3 etcdctl --endpoints=127.0.0.1:12379 lock my-terraform-lock -- watch date
```

This will sit on a loop printing the date every two seconds. If you start a second command against a different `etcd` member, you'll obviously be left waiting to acquire the lock:

```
$ ETCDCTL_API=3 etcdctl --endpoints=127.0.0.1:22379 lock my-terraform-lock -- watch date
```

Now, if you take down `127.0.0.1:12379` (the first member), the second command will still be left waiting to acquire the lock. If you take `127.0.0.1:12379` up again the first process keeps  printing the date. Only when the first process is stopped the second one takes over the lock and starts printing the date. Here's the code for the `lock` CLI command, which of course uses the same API as this implementation:

https://github.com/coreos/etcd/blob/master/etcdctl/ctlv3/command/lock_command.go

In the end I'd say we don't need to keep our own goroutine to handle keepalive/reconnect. WDYT @jbardin?

Reply:
I agree. Let's favor the simpler solution for now under the assumption that the etc client package takes care of all the details.